app/addons/documents/index-editor/actions.js

```diff
@@ -15,142 +15,282 @@ define([
+    // if the view name just changed and it's in the SAME design doc, remove the old one before saving the doc
+    if (viewInfo.originalDesignDocName === viewInfo.designDocId && viewInfo.originalViewName !== viewInfo.viewName) {
+      designDoc.removeDdocView(viewInfo.originalViewName);
+    }
 
+    designDoc.save().then(function () {
       FauxtonAPI.addNotification({
-        msg:  "Saving View...",
-        type: "info",
+        msg: 'View Saved.',
+        type: 'success',
         clear: true
       });
 
-      designDoc.save().then(function () {
-        FauxtonAPI.addNotification({
-          msg:  "View Saved.",
-          type: "success",
-          clear: true
-        });
+      // if the user just saved the view to a different design doc, remove the view from the old design doc and
+      // delete if it's empty
+      if (viewInfo.originalDesignDocName !== viewInfo.designDocId) {
+        var oldDesignDoc = findDesignDoc(viewInfo.designDocs, viewInfo.originalDesignDocName);
+        safeDeleteView(oldDesignDoc, viewInfo.designDocs, viewInfo.database, viewInfo.originalViewName);
+      }
 
-        if (_.any([viewInfo.designDocChanged, viewInfo.hasViewNameChanged, viewInfo.newDesignDoc, viewInfo.newView])) {
-          FauxtonAPI.dispatch({ type: ActionTypes.VIEW_SAVED });
-          var fragment = FauxtonAPI.urls('view', 'showNewlySavedView', viewInfo.database.safeID(), designDoc.safeID(), app.utils.safeURLName(viewInfo.viewName));
-          FauxtonAPI.navigate(fragment, { trigger: true });
-        } else {
-          this.updateDesignDoc(designDoc);
-        }
-
-        // this can be removed after the Views are on their own page
-        IndexResultsActions.reloadResultsList();
-      }.bind(this));
-    },
+      if (viewInfo.designDocId === 'new-doc') {
+        addDesignDoc(designDoc);
+      }
 
-    updateDesignDoc: function (designDoc) {
-      FauxtonAPI.dispatch({
-        type: ActionTypes.VIEW_UPDATE_DESIGN_DOC,
-        designDoc: designDoc.toJSON()
-      });
-    },
+      FauxtonAPI.dispatch({ type: ActionTypes.VIEW_SAVED });
+      var fragment = FauxtonAPI.urls('view', 'showView', viewInfo.database.safeID(), designDoc.safeID(), app.utils.safeURLName(viewInfo.viewName));
+      FauxtonAPI.navigate(fragment, { trigger: true });
+    });
+  }
 
-    deleteView: function (options) {
-      var viewName = options.viewName;
-      var database = options.database;
-      var designDoc = ActionHelpers.findDesignDoc(options.designDocs, options.designDocId);
-      var promise;
+  function addDesignDoc (designDoc) {
+    FauxtonAPI.dispatch({
+      type: ActionTypes.VIEW_ADD_DESIGN_DOC,
+      designDoc: designDoc.toJSON()
+    });
+  }
 
-      designDoc.removeDdocView(viewName);
+  function deleteView (options) {
 
-      if (designDoc.hasViews()) {
-        promise = designDoc.save();
-      } else {
-        promise = designDoc.destroy();
-      }
+    var onSuccess = function () {
```

Comment:
can we make this uniform and make this a named function, too?

both ways to write a function have different scoping implications on `this`, keeping them all "named" makes it easier to scan them